### ISSUE TITLE
The os_network module fails on creating a new neutron network in Cisco OpenStack cloud

### ISSUE THREAD

**Issue by bnaydenov** [ansible/ansible-modules-core#2435]
Issue Type:
Bug Report

Ansible Version:

ansible 2.0.0 (devel da4b4a2a33) last updated 2015/11/09 09:40:36 (GMT +300)
  lib/ansible/modules/core: (detached HEAD a3a7cdb9aa) last updated 2015/11/02 12:37:57 (GMT +300)
  lib/ansible/modules/extras: (detached HEAD b1db0e893c) last updated 2015/11/09 09:40:19 (GMT +300)

Ansible Configuration:

Environment:

N/A

Summary:

The os_network module fails on creating a new neutron network in Cisco OpenStack cloud. Account used is with super admin rights. I can successfully create new network from Horizon UI with the same account.

Steps To Reproduce:

```

---
- hosts: 127.0.0.1
  connection: local
  tasks:
    - name: create internal network
      os_network:
        name: "test-private"
        admin_state_up: yes
        shared: no
        external: no
        cloud: "test"
```

I also get the same result with "shared: yes" & "external: yes"

Expected Results:

I'd expect to be able to set new network named "test-private". 

Actual Results:

Currently I get:
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Error creating network test-private: Forbidden"}

**Comment by bcoca** [comment]
cc @emonty @Shrews

**Comment by Shrews** [comment]
@bnaydenov The error message being returned seems to indicate that the account you are using does not have the proper access rights. You will need to discuss this with your cloud admin. We do not have access to a Cisco cloud to verify and/or debug the behavior you are seeing.

We try our best to work on as many different providers as possible, but each sets up their cloud a bit differently, so you may have also come across an edge case we have yet to experience.

**Comment by emonty** [comment]
It's worth checking that the values in your clouds.yaml are correct - especially project and or domain values (not sure if cisco is doing keystone v3 and requires domains or not) Can you paste your clouds.yaml (please black out the actual password of course)

**Comment by bnaydenov** [comment]
@Shrews - with the same account via Horizon Web GUI I can create successfully new network. So assuming that account have not enough permission is wrong 

@emonty - I have double & triple check vars for project and domain. In fact I am using  proper OS_ environment variables  set. With this environment vars set i can successfully run openstack cli command for example: 

```
neutron net-list

```

I have even create a small python script with python shade and get same result, Error creating network test-private: Forbidden

Script exmaple:

```
#!/usr/bin/python
import shade

# I have right OS_ enviroment vars set 
cloud = shade.openstack_cloud()
try:
  net = cloud.create_network('test-private',external=False)
except Exception as e:
        print e.message (this prints "Error creating network test-private: Forbidden")
```

**what i have found so far and how i workaround this problem:**

I am using Ubuntu 14.04 with latest stable python shade version (shade (1.0.0))

with small hack in this file /usr/local/lib/python2.7/dist-packages/shade/**init**.py in my case line 1838 everything works correctly

```
def create_network(self, name, shared=False, admin_state_up=True,
                       external=False):
        """Create a network.

        :param string name: Name of the network being created.
        :param bool shared: Set the network as shared.
        :param bool admin_state_up: Set the network administrative state to up.
        :param bool external: Whether this network is externally accessible.

        :returns: The network object.
        :raises: OpenStackCloudException on operation error.
        """

        network = {
            'name': name,
            'shared': shared,
            'admin_state_up': admin_state_up,
           ###HACK####
            'router:external': external  !!!!Just delete this line!!!! 
           ###HACK####
        }

        with self._neutron_exceptions(
                "Error creating network {0}".format(name)):
            net = self.manager.submitTask(
                _tasks.NetworkCreate(body=dict({'network': network})))
        # Turns out neutron returns an actual dict, so no need for the
        # use of meta.obj_to_dict() here (which would not work against
        # a dict).
        return net['network']

```

When i delete this line " 'router:external': external" on line 1838 everything work correctly and I can create network with both ansible and test demo script above.

**Comment by emonty** [comment]
On 11/16/2015 11:51 AM, bogdan naydenov wrote:

> @Shrews https://github.com/Shrews - with the same account via Horizon
> Web GUI I can create successfully new network. So assuming that account
> have not enough permission is wrong
> 
> @emonty https://github.com/emonty - I have double & triple check vars
> for project and domain. In fact I am using proper OS_ environment
> variables set. With this environment vars set i can successfully run
> openstack cli command for example:
> 
> |neutron net-list
> 
> |
> |
> 
> I have event create a small python script with python and get same
> result, Error creating network test-private: Forbidden
> 
> Script exmaple:
> 
> |#!/usr/bin/python
> import shade
> 
> # I have right OS_ enviroment vars set
> 
> cloud = shade.openstack_cloud()
> try:
>    nett = cloud.create_network('test-private',external=False)
> except Exception as e:
>          print e.message (this prints "Error creating network test-private: Forbidden")
> |
> 
> _what i have found so far and how i workaround this problem:_
> 
> I am using Ubuntu 14.04 with latest stable python shade version (shade
> (1.0.0))
> 
> with small hack in this file
> /usr/local/lib/python2.7/dist-packages/shade/_init_.py everything works
> correctly
> 
> |def create_network(self, name, shared=False, admin_state_up=True,
>                         external=False):
>          """Create a network.
> 
> ```
>      :param string name: Name of the network being created.
>      :param bool shared: Set the network as shared.
>      :param bool admin_state_up: Set the network administrative state to up.
>      :param bool external: Whether this network is externally accessible.
> 
>      :returns: The network object.
>      :raises: OpenStackCloudException on operation error.
>      """
> 
>      network = {
>          'name': name,
>          'shared': shared,
>          'admin_state_up': admin_state_up,
>         ###HACK####
>          'router:external': external  !!!!Just delete this line!!!!
>         ###HACK####
>      }
> 
>      with self._neutron_exceptions(
>              "Error creating network {0}".format(name)):
>          net = self.manager.submitTask(
>              _tasks.NetworkCreate(body=dict({'network': network})))
>      # Turns out neutron returns an actual dict, so no need for the
>      # use of meta.obj_to_dict() here (which would not work against
>      # a dict).
>      return net['network']
> ```
> 
> |
> |
> 
> When i delete this line " 'router:external': external" everything work
> correctly and I can create network with both ansible and test demo
> script above.

Ah! Ok. That's helpful. I believe this is one of our favorite 
neutron-isms where not sending flags is sometimes important. I believe I 
know how to patch this to fix it for you.

**Comment by Shrews** [comment]
Newly released shade 1.2.0 does not send this neutron flag unless it is set to True.

**Comment by bnaydenov** [comment]
Hi, 

Just tested with shade 1.2 and everything works OK:-)))

Thanks for you help
